File: src_cpp/ai_factory/cpu/hull_white/caplets.cpp
#include "ai_factory/cpu/hull_white/caplets.hpp"
#include "ai_factory/common/fixed_income/hull_white.hpp"
#include "ai_factory/cpu/common/philox.hpp"
#include <algorithm>
#include <cmath>
#include <stdexcept>
namespace ai_factory::cpu::hull_white {
void price_caplet_batch(const cuda::HullWhiteCapletRow* rows,std::size_t count,std::size_t paths,cuda::MonteCarloOutput* outputs){
    if(paths<2U)throw std::invalid_argument("Caplet pricing requires at least two paths.");
#ifdef _OPENMP
#pragma omp parallel for schedule(static)
#endif
    for(std::ptrdiff_t i=0;i<static_cast<std::ptrdiff_t>(count);++i){const auto&r=rows[i];const auto&p=r.product;const auto normals=simulation::philox_standard_normals(r.seed,2U*paths,0U);const double state_var=fixed_income::hull_white_state_variance(r.mean_reversion,r.volatility,p.fixing_time);const double integral_var=fixed_income::hull_white_integral_variance(r.mean_reversion,r.volatility,p.fixing_time);const double covariance=fixed_income::hull_white_state_integral_covariance(r.mean_reversion,r.volatility,p.fixing_time);const double state_scale=std::sqrt(state_var);const double integral_loading=covariance/state_scale;const double residual_scale=std::sqrt(std::max(integral_var-integral_loading*integral_loading,0.0));const double deterministic_integral=fixed_income::hull_white_deterministic_integral(p.fixing_time,r.mean_reversion,r.volatility,r.beta0,r.beta1,r.beta2,r.tau);const double maturity=p.fixing_time+p.accrual_period;const double bond_a=fixed_income::hull_white_bond_a(p.fixing_time,maturity,r.mean_reversion,r.volatility,r.beta0,r.beta1,r.beta2,r.tau);const double bond_b=fixed_income::hull_white_b(r.mean_reversion,p.accrual_period);double sum=0.0,sumsq=0.0;for(std::size_t path=0;path<paths;++path){const double z0=normals[2U*path],z1=normals[2U*path+1U];const double state=state_scale*z0;const double integral=integral_loading*z0+residual_scale*z1;const double bond=bond_a*std::exp(-bond_b*state);const double payoff=std::exp(-deterministic_integral-integral)*p.notional*std::max(1.0-(1.0+p.accrual_period*p.strike)*bond,0.0);sum+=payoff;sumsq+=payoff*payoff;}const double n=static_cast<double>(paths),mean=sum/n,var=(sumsq-n*mean*mean)/(n-1.0);outputs[i]={mean,std::sqrt(std::max(var,0.0)/n)};}
}
}

File: src_python/ai_factory/pytorch/hull_white/caplets.py
"""Batched exact-transition Hull-White caplet Monte Carlo."""

from __future__ import annotations

from time import perf_counter
from typing import Any

import torch

from ai_factory.pytorch.common.device import resolve_device, synchronize, warmup
from ai_factory.pytorch.common.fixed_income import (
    hull_white_bond_coefficients,
    hull_white_integral_variance,
    hull_white_state_integral_covariance,
    hull_white_state_variance,
    nelson_siegel_discount,
)
from ai_factory.pytorch.common.monte_carlo import price_summary

DEFAULT_BATCH_ROWS = 128


def _column(values: list[float], target: torch.device) -> torch.Tensor:
    return torch.tensor(values, device=target, dtype=torch.float64)[:, None]


def _price_batch_impl(rows, models, curves, products, *, num_paths, target, batch_rows):
    started = perf_counter()
    outputs: list[dict[str, float]] = []
    for offset in range(0, len(rows), batch_rows):
        batch = rows[offset : offset + batch_rows]
        ms = [models[row["model_id"]] for row in batch]
        cs = [curves[row["curve_id"]] for row in batch]
        ps = [products[row["product_id"]] for row in batch]
        a = _column([model["mean_reversion"] for model in ms], target)
        sigma = _column([model["volatility"] for model in ms], target)
        beta0 = _column([curve["beta0"] for curve in cs], target)
        beta1 = _column([curve["beta1"] for curve in cs], target)
        beta2 = _column([curve["beta2"] for curve in cs], target)
        tau = _column([curve["tau"] for curve in cs], target)
        fixing = _column([product["fixing_time"] for product in ps], target)
        accrual = _column([product["accrual_period"] for product in ps], target)
        strike = _column([product["strike"] for product in ps], target)
        notional = _column([product["notional"] for product in ps], target)

        bond_a, bond_b = hull_white_bond_coefficients(
            fixing, fixing + accrual, a, sigma, beta0, beta1, beta2, tau
        )
        normals = torch.randn(
            (len(batch), num_paths, 2), device=target, dtype=torch.float64
        )
        state_variance = hull_white_state_variance(a, sigma, fixing)
        integral_variance = hull_white_integral_variance(a, sigma, fixing)
        covariance = hull_white_state_integral_covariance(a, sigma, fixing)
        state_scale = torch.sqrt(state_variance)
        integral_loading = covariance / state_scale
        state = state_scale * normals[:, :, 0]
        state_integral = integral_loading * normals[:, :, 0] + torch.sqrt(
            torch.clamp(integral_variance - integral_loading.square(), min=0.0)
        ) * normals[:, :, 1]
        discount = torch.exp(
            torch.log(nelson_siegel_discount(fixing, beta0, beta1, beta2, tau))
            - 0.5 * integral_variance
            - state_integral
        )
        bond = bond_a * torch.exp(-bond_b * state)
        payoff = discount * notional * torch.clamp(
            1.0 - (1.0 + accrual * strike) * bond, min=0.0
        )
        outputs.extend(price_summary(payoff))
    synchronize(target)
    return outputs, {"wall_seconds": perf_counter() - started}


def price_batch(
    rows: list[dict[str, Any]],
    model_by_id: dict[str, Any],
    curve_by_id: dict[str, Any],
    product_by_id: dict[str, Any],
    *,
    num_paths: int,
    device: str,
    batch_rows: int = DEFAULT_BATCH_ROWS,
    **_: Any,
):
    target = resolve_device(device)
    if target.type == "cuda" and rows:
        warmup(target, shape=(min(batch_rows, len(rows)), num_paths, 2), dtype=torch.float64)
        _price_batch_impl(
            rows[:batch_rows], model_by_id, curve_by_id, product_by_id,
            num_paths=num_paths, target=target, batch_rows=batch_rows,
        )
    return _price_batch_impl(
        rows, model_by_id, curve_by_id, product_by_id,
        num_paths=num_paths, target=target, batch_rows=batch_rows,
    )
